File: server/internal/protocol/message.go
package protocol

import "encoding/json"

// Envelope 是所有消息的统一信封。
type Envelope struct {
	T    string          `json:"t"`              // 消息类型
	ID   string          `json:"id,omitempty"`   // 请求 ID，用于 req/resp 匹配；推送/同步可空
	Data json.RawMessage `json:"data,omitempty"` // 消息体，按 T 解析
}

// 消息类型常量
const (
	TypeLogin         = "login"            // 请求：登录
	TypeSync          = "sync"             // 推送：全量同步
	TypeLoot          = "loot"             // 推送：掉落装备
	TypeFloor         = "floor"            // 推送：层数推进
	TypeEquip         = "equip"            // 请求：穿戴
	TypeUnequip       = "unequip"          // 请求：卸下
	TypeBag           = "bag"              // 推送：背包全量
	TypePower         = "power"            // 推送：当前战力
	TypeCombat        = "combat"           // 推送：战斗 tick 结果
	TypeDecompose     = "decompose"        // 请求：分解
	TypeCompose       = "compose"          // 请求：合成
	TypeReforge       = "reforge"          // 请求：重铸
	TypeUpgrade       = "upgrade"          // 请求：强化
	TypeTransferUpg   = "transfer_upgrade" // 请求：强化继承/转移
	TypeLockEquipment = "lock_equipment"   // 请求：锁定/解锁装备
	TypeMaterials     = "materials"        // 推送：材料库存
	TypeCraftResult   = "craft_result"     // 推送：养成操作结果
	TypeOfflineResult = "offline_result"   // 推送：离线结算结果
	TypeReincarn      = "reincarn"         // 请求：转生
	TypeTalentUp      = "talent_up"        // 请求：天赋升级
	TypeTalents       = "talents"          // 推送：天赋状态
)

// LoginRequest 登录请求体。
type LoginRequest struct {
	Account string `json:"account"` // 临时账号名
}

// SyncData 全量同步消息体。阶段 0 最小字段，后续阶段扩充。
type SyncData struct {
	Account              string   `json:"account"`                // 账号
	Floor                int      `json:"floor"`                  // 当前层数
	FloorKills           int      `json:"floor_kills"`            // 当前层小兵击杀数
	MinionsTotal         int      `json:"minions_total"`          // 每层小兵总数
	EquipmentDataVersion int      `json:"equipment_data_version"` // 装备基底数据版本
	LegendaryDataVersion int      `json:"legendary_data_version"` // 传奇定义数据版本
	ArtifactDataVersion  int      `json:"artifact_data_version"`  // 神器定义数据版本
	Souls                int      `json:"souls"`                  // 魂点
	Inventory            []string `json:"inventory"`              // 背包物品 ID 占位
}

// CombatData describes one authoritative combat tick for client animation.
type CombatData struct {
	Floor             int              `json:"floor"`
	EnemyKind         string           `json:"enemy_kind"`
	EnemyFamily       string           `json:"enemy_family,omitempty"`
	EnemyElement      string           `json:"enemy_element,omitempty"`
	Win               bool             `json:"win"`
	PlayerPower       float64          `json:"player_power"`
	EnemyPower        float64          `json:"enemy_power"`
	EnemyResistances  []ResistanceDTO  `json:"enemy_resistances,omitempty"`
	MinionsKilled     int              `json:"minions_killed"`
	MinionsTotal      int              `json:"minions_total"`
	FloorAdvanced     bool             `json:"floor_advanced"`
	PlayerStartHP     float64          `json:"player_start_hp,omitempty"`
	EnemyStartHP      float64          `json:"enemy_start_hp,omitempty"`
	PlayerStartShield float64          `json:"player_start_shield,omitempty"`
	EnemyStartShield  float64          `json:"enemy_start_shield,omitempty"`
	PlayerEndHP       float64          `json:"player_end_hp,omitempty"`
	EnemyEndHP        float64          `json:"enemy_end_hp,omitempty"`
	PlayerEndShield   float64          `json:"player_end_shield,omitempty"`
	EnemyEndShield    float64          `json:"enemy_end_shield,omitempty"`
	Events            []CombatEventDTO `json:"events,omitempty"`
}

// ResistanceDTO describes one elemental resistance or vulnerability.
type ResistanceDTO struct {
	Type  string  `json:"type"`
	Value float64 `json:"value"`
}

// CombatEventDTO is one server-authored animation event inside a combat tick.
type CombatEventDTO struct {
	Index        int     `json:"index"`
	Actor        string  `json:"actor"`
	Kind         string  `json:"kind"`
	Damage       float64 `json:"damage,omitempty"`
	Critical     bool    `json:"critical,omitempty"`
	PlayerHP     float64 `json:"player_hp"`
	EnemyHP      float64 `json:"enemy_hp"`
	PlayerShield float64 `json:"player_shield,omitempty"`
	EnemyShield  float64 `json:"enemy_shield,omitempty"`
}

// LootData 掉落推送消息体。
type LootData struct {
	UID                  string     `json:"uid"`     // 装备唯一 ID
	BaseID               string     `json:"base_id"` // 基底 ID
	LegendaryID          string     `json:"legendary_id,omitempty"`
	ArtifactID           string     `json:"artifact_id,omitempty"`
	LegendaryDescription string     `json:"legendary_description,omitempty"`
	ArtifactDescription  string     `json:"artifact_description,omitempty"`
	LegendaryBonuses     []AffixDTO `json:"legendary_bonuses,omitempty"`
	ArtifactBonuses      []AffixDTO `json:"artifact_bonuses,omitempty"`
	LegendaryPowerBonus  float64    `json:"legendary_power_bonus,omitempty"`
	BossRewardBonus      float64    `json:"boss_reward_bonus,omitempty"`
	ArtifactTrigger      string     `json:"artifact_trigger,omitempty"`
	ArtifactValue        float64    `json:"artifact_value,omitempty"`
	Name                 string     `json:"name"`    // 装备名
	Slot                 int        `json:"slot"`    // 槽位
	Rarity               int        `json:"rarity"`  // 稀有度
	Upgrade              int        `json:"upgrade"` // 强化等级
	Affixes              []AffixDTO `json:"affixes"` // 词缀列表
}

// AffixDTO 词缀传输对象。
type AffixDTO struct {
	Type  string  `json:"type"`
	Tier  int     `json:"tier"`
	Value float64 `json:"value"`
}

// FloorData 层数推送消息体。
type FloorData struct {
	Floor int `json:"floor"` // 新层数
}

// EquipRequest 穿戴请求体。
type EquipRequest struct {
	UID string `json:"uid"`
}

// UnequipRequest 卸下请求体。
type UnequipRequest struct {
	Slot int `json:"slot"`
}

// BagData 背包全量推送。
type BagData struct {
	Items    []EquipmentDTO `json:"items"`
	Equipped []EquipmentDTO `json:"equipped"`
}

// EquipmentDTO 装备传输对象（背包与已穿戴通用）。
type EquipmentDTO struct {
	UID                  string     `json:"uid"`
	BaseID               string     `json:"base_id"`
	LegendaryID          string     `json:"legendary_id,omitempty"`
	ArtifactID           string     `json:"artifact_id,omitempty"`
	LegendaryDescription string     `json:"legendary_description,omitempty"`
	ArtifactDescription  string     `json:"artifact_description,omitempty"`
	LegendaryBonuses     []AffixDTO `json:"legendary_bonuses,omitempty"`
	ArtifactBonuses      []AffixDTO `json:"artifact_bonuses,omitempty"`
	LegendaryPowerBonus  float64    `json:"legendary_power_bonus,omitempty"`
	BossRewardBonus      float64    `json:"boss_reward_bonus,omitempty"`
	ArtifactTrigger      string     `json:"artifact_trigger,omitempty"`
	ArtifactValue        float64    `json:"artifact_value,omitempty"`
	Name                 string     `json:"name"`
	Slot                 int        `json:"slot"`
	Rarity               int        `json:"rarity"`
	Upgrade              int        `json:"upgrade"`
	Locked               bool       `json:"locked"`
	PowerScore           float64    `json:"power_score"`
	PowerScoreValid      bool       `json:"power_score_valid"`
	Affixes              []AffixDTO `json:"affixes"`
}

// PowerData 战力推送。
type PowerData struct {
	Power float64 `json:"power"`
}

// DecomposeRequest 分解请求体。
type DecomposeRequest struct {
	UID string `json:"uid"`
}

// ComposeRequest 合成请求体。
type ComposeRequest struct {
	Slot int `json:"slot"`
}

// ReforgeRequest 重铸请求体。
type ReforgeRequest struct {
	UID string `json:"uid"`
}

// UpgradeRequest 强化请求体。
type UpgradeRequest struct {
	UID string `json:"uid"`
}

// TransferUpgradeRequest 强化继承请求体。
type TransferUpgradeRequest struct {
	SourceUID string `json:"source_uid"`
	TargetUID string `json:"target_uid"`
}

// LockEquipmentRequest 锁定/解锁装备请求体。
type LockEquipmentRequest struct {
	UID    string `json:"uid"`
	Locked bool   `json:"locked"`
}

// MaterialKV 材料键值对。
type MaterialKV struct {
	Key string `json:"k"`
	Val int    `json:"v"`
}

// MaterialsData 材料库存推送（数组格式，客户端 JsonUtility 可解析）。
type MaterialsData struct {
	Materials []MaterialKV `json:"materials"`
}

// TalentKV 天赋键值对。
type TalentKV struct {
	Name  string `json:"name"`
	Level int    `json:"level"`
}

// TalentsData 天赋状态推送（数组格式，客户端 JsonUtility 可解析）。
type TalentsData struct {
	Souls       int        `json:"souls"`
	MaxFloor    int        `json:"max_floor"`
	CanReincarn bool       `json:"can_reincarn"`
	Talents     []TalentKV `json:"talents"`
}

// CraftResult 养成操作结果（分解/合成/重铸/强化通用响应推送）。
type CraftResult struct {
	OK      bool   `json:"ok"`
	Msg     string `json:"msg"`
	UID     string `json:"uid,omitempty"`
	Upgrade int    `json:"upgrade,omitempty"`
}

// OfflineResultData 离线结算结果推送。
type OfflineResultData struct {
	DurationSeconds int `json:"duration_seconds"` // 结算时长（秒）
	TicksSimulated  int `json:"ticks_simulated"`  // 模拟 tick 数
	LootCount       int `json:"loot_count"`       // 掉落数
	FloorsAdvanced  int `json:"floors_advanced"`  // 推进层数
}

// TalentUpRequest 天赋升级请求体。
type TalentUpRequest struct {
	Name string `json:"name"`
}

File: client/Assets/Scripts/Net/Message.cs
using System;

namespace EquipmentIdle.Net
{
    /// <summary>
    /// WebSocket 消息信封编解码。
    /// 信封格式：{ "t": "<type>", "id": "<reqId>", "data": {...} }
    /// 使用 Unity 内置 JsonUtility，零第三方依赖。
    /// 注意：JsonUtility 要求顶层对象可序列化，data 用具体类型类承载。
    /// </summary>
    public static class Message
    {
        public const string TypeLogin = "login";
        public const string TypeSync = "sync";
        public const string TypeLoot = "loot";
        public const string TypeFloor = "floor";
        public const string TypeEquip = "equip";
        public const string TypeUnequip = "unequip";
        public const string TypeBag = "bag";
        public const string TypePower = "power";
        public const string TypeCombat = "combat";
        public const string TypeDecompose = "decompose";
        public const string TypeCompose = "compose";
        public const string TypeReforge = "reforge";
        public const string TypeUpgrade = "upgrade";
        public const string TypeTransferUpgrade = "transfer_upgrade";
        public const string TypeLockEquipment = "lock_equipment";
        public const string TypeMaterials = "materials";
        public const string TypeCraftResult = "craft_result";
        public const string TypeReincarn = "reincarn";
        public const string TypeTalentUp = "talent_up";
        public const string TypeTalents = "talents";
        public const string TypeOfflineResult = "offline_result";

        /// <summary>编码登录请求为信封 JSON 字符串。</summary>
        public static string EncodeLogin(string id, string account)
        {
            string dataJson = "{\"account\":\"" + Escape(account) + "\"}";
            return "{\"t\":\"" + TypeLogin + "\",\"id\":\"" + Escape(id) + "\",\"data\":" + dataJson + "}";
        }

        /// <summary>编码穿戴请求。</summary>
        public static string EncodeEquip(string id, string uid)
        {
            string dataJson = "{\"uid\":\"" + Escape(uid) + "\"}";
            return "{\"t\":\"" + TypeEquip + "\",\"id\":\"" + Escape(id) + "\",\"data\":" + dataJson + "}";
        }

        /// <summary>编码卸下请求。</summary>
        public static string EncodeUnequip(string id, int slot)
        {
            string dataJson = "{\"slot\":" + slot + "}";
            return "{\"t\":\"" + TypeUnequip + "\",\"id\":\"" + Escape(id) + "\",\"data\":" + dataJson + "}";
        }

        /// <summary>编码分解请求。</summary>
        public static string EncodeDecompose(string id, string uid)
        {
            string dataJson = "{\"uid\":\"" + Escape(uid) + "\"}";
            return "{\"t\":\"" + TypeDecompose + "\",\"id\":\"" + Escape(id) + "\",\"data\":" + dataJson + "}";
        }

        /// <summary>编码合成请求。</summary>
        public static string EncodeCompose(string id, int slot)
        {
            string dataJson = "{\"slot\":" + slot + "}";
            return "{\"t\":\"" + TypeCompose + "\",\"id\":\"" + Escape(id) + "\",\"data\":" + dataJson + "}";
        }

        /// <summary>编码重铸请求。</summary>
        public static string EncodeReforge(string id, string uid)
        {
            string dataJson = "{\"uid\":\"" + Escape(uid) + "\"}";
            return "{\"t\":\"" + TypeReforge + "\",\"id\":\"" + Escape(id) + "\",\"data\":" + dataJson + "}";
        }

        /// <summary>编码强化请求。</summary>
        public static string EncodeUpgrade(string id, string uid)
        {
            string dataJson = "{\"uid\":\"" + Escape(uid) + "\"}";
            return "{\"t\":\"" + TypeUpgrade + "\",\"id\":\"" + Escape(id) + "\",\"data\":" + dataJson + "}";
        }

        /// <summary>编码强化继承请求。</summary>
        public static string EncodeTransferUpgrade(string id, string sourceUid, string targetUid)
        {
            string dataJson = "{\"source_uid\":\"" + Escape(sourceUid) + "\",\"target_uid\":\"" + Escape(targetUid) + "\"}";
            return "{\"t\":\"" + TypeTransferUpgrade + "\",\"id\":\"" + Escape(id) + "\",\"data\":" + dataJson + "}";
        }

        /// <summary>编码锁定/解锁请求。</summary>
        public static string EncodeLockEquipment(string id, string uid, bool locked)
        {
            string dataJson = "{\"uid\":\"" + Escape(uid) + "\",\"locked\":" + (locked ? "true" : "false") + "}";
            return "{\"t\":\"" + TypeLockEquipment + "\",\"id\":\"" + Escape(id) + "\",\"data\":" + dataJson + "}";
        }

        /// <summary>编码转生请求。</summary>
        public static string EncodeReincarn(string id)
        {
            return "{\"t\":\"" + TypeReincarn + "\",\"id\":\"" + Escape(id) + "\"}";
        }

        /// <summary>编码天赋升级请求。</summary>
        public static string EncodeTalentUp(string id, string name)
        {
            string dataJson = "{\"name\":\"" + Escape(name) + "\"}";
            return "{\"t\":\"" + TypeTalentUp + "\",\"id\":\"" + Escape(id) + "\",\"data\":" + dataJson + "}";
        }

        /// <summary>将材料数组转为字典。</summary>
        public static System.Collections.Generic.Dictionary<string, int> MaterialsToDict(MaterialEntry[] entries)
        {
            var dict = new System.Collections.Generic.Dictionary<string, int>();
            if (entries == null) return dict;
            foreach (var e in entries) dict[e.k] = e.v;
            return dict;
        }

        /// <summary>将天赋数组转为字典。</summary>
        public static System.Collections.Generic.Dictionary<string, int> TalentsToDict(TalentEntry[] entries)
        {
            var dict = new System.Collections.Generic.Dictionary<string, int>();
            if (entries == null) return dict;
            foreach (var e in entries) dict[e.name] = e.level;
            return dict;
        }

        /// <summary>解析收到的信封文本。失败返回 null。</summary>
        public static ParsedMessage Parse(string text)
        {
            try
            {
                // JsonUtility 无法解析顶层含 data 为对象的混合结构，
                // 这里用轻量手写提取 t / id / data 三段。
                var pm = new ParsedMessage();
                pm.t = ExtractString(text, "t");
                pm.id = ExtractString(text, "id");

                // data 子对象：找到 "data": 之后到信封结尾的匹配大括号
                pm.dataJson = ExtractObject(text, "data");
                if (pm.t == null) return null;
                return pm;
            }
            catch
            {
                return null;
            }
        }

        /// <summary>转义 JSON 字符串里的特殊字符。</summary>
        private static string Escape(string s)
        {
            if (string.IsNullOrEmpty(s)) return "";
            return s.Replace("\\", "\\\\").Replace("\"", "\\\"");
        }

        /// <summary>从 JSON 文本里提取某个字符串字段的值（仅匹配 "key":"value" 形式）。</summary>
        private static string ExtractString(string json, string key)
        {
            string pattern = "\"" + key + "\":\"";
            int i = json.IndexOf(pattern, StringComparison.Ordinal);
            if (i < 0) return "";
            i += pattern.Length;
            int end = json.IndexOf('"', i);
            if (end < 0) return "";
            return json.Substring(i, end - i);
        }

        /// <summary>从 JSON 文本里提取某个对象字段的 JSON 子串（匹配 "key":{...}）。</summary>
        private static string ExtractObject(string json, string key)
        {
            string pattern = "\"" + key + "\":";
            int i = json.IndexOf(pattern, StringComparison.Ordinal);
            if (i < 0) return "{}";
            i += pattern.Length;
            // 跳过空白
            while (i < json.Length && (json[i] == ' ' || json[i] == '\t')) i++;
            if (i >= json.Length || json[i] != '{') return "{}";
            int depth = 0;
            int start = i;
            for (; i < json.Length; i++)
            {
                if (json[i] == '{') depth++;
                else if (json[i] == '}')
                {
                    depth--;
                    if (depth == 0) return json.Substring(start, i - start + 1);
                }
            }
            return "{}";
        }
    }

    /// <summary>解析后的消息。dataJson 是 data 字段的原始 JSON 子串，供具体类型反序列化。</summary>
    public class ParsedMessage
    {
        public string t;
        public string id;
        public string dataJson;
    }

    /// <summary>sync 消息的 data 结构，与服务端 protocol.SyncData 对应。</summary>
    [Serializable]
    public class SyncData
    {
        public string account;
        public int floor;
        public int floor_kills;
        public int minions_total;
        public int equipment_data_version;
        public int legendary_data_version;
        public int artifact_data_version;
        public int souls;
        public string[] inventory;
    }

    /// <summary>服务端权威战斗 tick，用于驱动关卡进度和客户端动画。</summary>
    [Serializable]
    public class CombatData
    {
        public int floor;
        public string enemy_kind;
        public string enemy_family;
        public string enemy_element;
        public bool win;
        public float player_power;
        public float enemy_power;
        public ResistanceData[] enemy_resistances;
        public int minions_killed;
        public int minions_total;
        public bool floor_advanced;
        public float player_start_hp;
        public float enemy_start_hp;
        public float player_start_shield;
        public float enemy_start_shield;
        public float player_end_hp;
        public float enemy_end_hp;
        public float player_end_shield;
        public float enemy_end_shield;
        public CombatEventData[] events;
    }

    [Serializable]
    public class CombatEventData
    {
        public int index;
        public string actor;
        public string kind;
        public float damage;
        public bool critical;
        public float player_hp;
        public float enemy_hp;
        public float player_shield;
        public float enemy_shield;
    }

    [Serializable]
    public class ResistanceData
    {
        public string type;
        public float value;
    }

    /// <summary>login 消息的 data 结构。</summary>
    [Serializable]
    public class LoginData
    {
        public string account;
    }

    /// <summary>掉落推送的 data 结构。</summary>
    [Serializable]
    public class LootData
    {
        public string uid;
        public string base_id;
        public string legendary_id;
        public string artifact_id;
        public string legendary_description;
        public string artifact_description;
        public AffixData[] legendary_bonuses;
        public AffixData[] artifact_bonuses;
        public float legendary_power_bonus;
        public float boss_reward_bonus;
        public string artifact_trigger;
        public float artifact_value;
        public string name;
        public int slot;
        public int rarity;
        public int upgrade;
        public bool locked;
        public AffixData[] affixes;
    }

    /// <summary>层数推送的 data 结构。</summary>
    [Serializable]
    public class FloorData
    {
        public int floor;
    }

    /// <summary>装备传输对象（背包与已穿戴通用）。</summary>
    [Serializable]
    public class EquipmentDTO
    {
        public string uid;
        public string base_id;
        public string legendary_id;
        public string artifact_id;
        public string legendary_description;
        public string artifact_description;
        public AffixData[] legendary_bonuses;
        public AffixData[] artifact_bonuses;
        public float legendary_power_bonus;
        public float boss_reward_bonus;
        public string artifact_trigger;
        public float artifact_value;
        public string name;
        public int slot;
        public int rarity;
        public int upgrade;
        public bool locked;
        public float power_score;
        public bool power_score_valid;
        public AffixData[] affixes;
    }

    /// <summary>词缀传输对象。</summary>
    [Serializable]
    public class AffixData
    {
        public string type;
        public int tier;
        public float value;
    }

    /// <summary>背包全量推送。</summary>
    [Serializable]
    public class BagData
    {
        public EquipmentDTO[] items;
        public EquipmentDTO[] equipped;
    }

    /// <summary>战力推送。</summary>
    [Serializable]
    public class PowerData
    {
        public float power;
    }

    /// <summary>养成操作结果推送。</summary>
    [Serializable]
    public class CraftResultData
    {
        public bool ok;
        public string msg;
        public string uid;
        public int upgrade;
    }

    /// <summary>材料库存推送（数组格式）。</summary>
    [Serializable]
    public class MaterialsData
    {
        public MaterialEntry[] materials;
    }

    /// <summary>材料键值对。</summary>
    [Serializable]
    public class MaterialEntry
    {
        public string k;
        public int v;
    }

    /// <summary>天赋状态推送（数组格式）。</summary>
    [Serializable]
    public class TalentsData
    {
        public int souls;
        public int max_floor;
        public bool can_reincarn;
        public TalentEntry[] talents;
    }

    /// <summary>天赋键值对。</summary>
    [Serializable]
    public class TalentEntry
    {
        public string name;
        public int level;
    }

    /// <summary>离线结算结果推送。</summary>
    [Serializable]
    public class OfflineResultData
    {
        public int duration_seconds;
        public int ticks_simulated;
        public int loot_count;
        public int floors_advanced;
    }
}
